src/plrust_proc.rs

```diff
@@ -0,0 +1,119 @@
+//! Routines for managing the `plrust.plrust_proc` extension table along with the data it contains
+use crate::error::PlRustError;
+use crate::gucs;
+use crate::user_crate::{StateLoaded, UserCrate};
+use pgx::pg_sys::MyDatabaseId;
+use pgx::{extension_sql, pg_sys, IntoDatum, PgBuiltInOids, PgOid, Spi};
+use std::path::Path;
+
+extension_sql!(
+    r#"
+CREATE TABLE plrust.plrust_proc (
+    id            regproc   NOT NULL,
```

Comment:
Ah I remember why I didn't use regproc and decided to go with object identity:

https://www.postgresql.org/docs/current/pgupgrade.html

pg_upgrade doesn't allow you to upgrade with regproc. 

Which also brings up an interesting question of how does this work with pg_upgrade...

If you use something like *link* mode, the data files including this table will be copied over? But CREATE OR REPLACE FUNCTION would run as part of pg_restore iirc, so you'll have stale entries in there presumably since I don't think functionOids are preserved right now

Reply:
I'll have to think about this, but right now my gut reaction is that this is pg_upgrade's problem, not pl/rust's.

Reply:
Yea but if we don't support it, I think it makes PL/Rust un-useable by anyone who cares about migrating to latest versions of Postgres which is going to be almost everyone eventually.

I'm skeptical the community will relax it's restrictions on regproc anytime soon.

It'll be a hindrance to adoption, and more of an unpleasant surprise when folks try upgrading since I doubt anyone actually reads the documentation properly.

IIUC you're just looking for a unique identity which you should be able to get by just using:

https://www.postgresql.org/docs/9.4/functions-info.html
`pg_identify_object`

I think it's a bit more of a hassle if the function schema migrates I guess but that should be able to be handled

Reply:
As I understand it, pg_upgrade will migrate our table with the existing regproc values.  Then at some point it'll CREATE OR REPLACE all user functions.  Then our table will have the old values, which are now likely stale, and the entries for the new CREATE OR REPLACE execution.

Other than some garbage in our table, I don't see the problem.  We can expose a function to garbage collect it -- it's a simple DELETE statement.

Reply:
It won't even try migrating. It checks if regproc is being used and blocks the upgrade. 
https://github.com/postgres/postgres/blob/master/src/bin/pg_upgrade/check.c#L1273

```
pg_upgrade does not support upgrading of databases containing table columns using these reg* OID-referencing system data types:
```

https://www.postgresql.org/docs/current/pgupgrade.html

Here's a sample repro I just did


```
postgres=> create table test (id regproc);
CREATE TABLE
postgres=> insert into test values (16405);
INSERT 0 1
postgres=> select * from test;
    id     
-----------
 increment
 ````
 
 Trigger pg_upgrade....
 
```
Your installation contains one of the reg* data types in user tables.
These data types reference system OIDs that are not preserved by
pg_upgrade, so this cluster cannot currently be upgraded.  You can
drop the problem columns and restart the upgrade.
A list of the problem columns is in the file:
    tables_using_reg.txt
```

I agree the case you're doing *could* be supported in theory since the handler handles all of that but it just blocks cause funcOid's aren't preserved so for the sake of data integrity they just prevent it.

I actually don't know if this scheme even will work with pg_upgrade if you're putting it in a table now that I think about it....

https://github.com/postgres/postgres/blob/master/src/bin/pg_upgrade/pg_upgrade.c#L159
https://github.com/postgres/postgres/blob/master/src/bin/pg_upgrade/pg_upgrade.c#L169-L172

We run pg_restore in `create_new_objects` which I'm not sure if it'll recreate the extension table at that step or not. And if it does and you run link it would break. 

I think that's more reason to use objectIdentity actually, cause in CREATE FUNCTION, you could just avoid recompiling if you know this is during pg_upgrade and that would be faster and you would get database link for free without needing to redo anything

Reply:
You're harshing my mellow, man.

I'll take a look at this today/tomorrow and put up a PR.  It won't be a big deal to use identify_object.  Will just need to make sure it's out of the "hot path" for execution.

Pretty annoying pg_upgrade can't handle this.

Reply:
> You're harshing my mellow, man.

Having to also deal with the function schema changing is annoying too.  We already have the hook in place so the hard work has been done, it's just now more complexity.

Reply:
> Having to also deal with the function schema changing is annoying too

Which, of course, also means on rename of a schema that contains plrust functions.

Geez, anything else I'm not thinking of?

Reply:
No I think that's about it. Off the top of my head it should just be:

1. Prevent schema from being renamed
2. Prevent direct table access for INSERT/DELETE? (SELECT seems fine)